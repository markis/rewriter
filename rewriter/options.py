from argparse import ArgumentParser
from typing import Protocol


class Options(Protocol):
    filename: str
    source: str
    dry_run: bool
    verbose: bool


def parse_options() -> Options:
    parser = ArgumentParser(prog="rewriter", description="Rewrite python code")
    parser.add_argument("filename")
    parser.add_argument("-d", "--dry-run", action="store_true")
    parser.add_argument("-v", "--verbose", action="store_true")

    opts: Options = parser.parse_args()
    return opts
