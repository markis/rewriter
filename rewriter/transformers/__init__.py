import pkgutil
from ast import AST
from collections import defaultdict
from collections.abc import Callable, Iterable

from rewriter.trackers.changes import Change
from rewriter.trackers.imports import Import

TransformerReturnType = tuple[Iterable[Change], Iterable[Import]]
Transformer = Callable[[AST, AST, AST], TransformerReturnType]
Transform = Callable[[AST, AST, AST], Iterable[TransformerReturnType]]

__all__ = ["Transform", "TransformerReturnType", "transform", "transformer"]

__registry = defaultdict(list)


def transformer(*types: type[AST]) -> Callable[[Transformer], None]:
    """
    Decorator to register a function as a transformer
    """

    def wrapper(func: Transformer) -> None:
        for t in types:
            __registry[t].append(func)

    return wrapper


def transform(node: AST, parent: AST, ctx: AST) -> Iterable[TransformerReturnType]:
    """
    Transform a node using the transformers in the registry
    """
    return [func(node, parent, ctx) for func in __registry[type(node)]]


def load_transformers(base_dir: str) -> None:
    """
    Automatically load every module in this folder, which will automatically register transformers
    """

    for _, module, __ in pkgutil.iter_modules([f"{base_dir}/transformers/"]):
        del _, __
        __import__(f"rewriter.transformers.{module}")
