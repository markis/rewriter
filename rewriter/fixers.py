import ast
from abc import ABC, abstractmethod
from collections import defaultdict
from collections.abc import Mapping, Sequence

from rewriter.options import Options
from rewriter.trackers.imports import ImportTracker
from rewriter.trackers.stats import ChangeTracker

ASTType = type[ast.AST]
FixTypes = Sequence[ASTType]
FixerMap = Mapping[ASTType, Sequence["Fixer"]]


NAME_ANY = "Any"
NAME_BOOL = "bool"
NAME_INT = "int"
NAME_NONE = "None"
NAME_STR = "str"
NAME_CLS = "cls"
NAME_SELF = "self"
NAME_TYPING = "typing"


class Fixer(ABC):
    opts: Options
    import_tracker: ImportTracker
    change_tracker: ChangeTracker

    def __init__(
        self, opts: Options, stats_tracker: ChangeTracker, import_tracker: ImportTracker
    ) -> None:
        self.opts = opts
        self.import_tracker = import_tracker
        self.change_tracker = stats_tracker

    @classmethod
    def get_fixers(
        cls, opts: Options, stats_tracker: ChangeTracker, import_tracker: ImportTracker
    ) -> FixerMap:
        fixers = defaultdict(list)
        for fixer in cls.__subclasses__():
            for fixer_type in fixer.get_fix_types():
                fixers[fixer_type].append(fixer(opts, stats_tracker, import_tracker))
        return fixers

    @classmethod
    @abstractmethod
    def get_fix_types(cls) -> FixTypes:
        raise NotImplementedError()

    @abstractmethod
    def fix(self, node: ast.AST, parent: ast.AST, ctx: ast.AST) -> None:
        raise NotImplementedError()


class ArgumentFixer(Fixer):
    @classmethod
    def get_fix_types(cls) -> FixTypes:
        return [ast.arguments]

    def fix(self, node: ast.AST, parent: ast.AST, ctx: ast.AST) -> None:
        if isinstance(node, ast.arguments):
            self.fix_args(node, parent, ctx)

    def fix_args(self, node: ast.arguments, parent: ast.AST, ctx: ast.AST) -> None:
        if node.vararg:
            self.fix_arg(node.vararg, parent, ctx)

        if node.kwarg:
            self.fix_arg(node.kwarg, parent, ctx)

        for arg in node.args:
            self.fix_arg(arg, parent, ctx)

        for arg in node.posonlyargs:
            self.fix_arg(arg, parent, ctx)

        for arg in node.kwonlyargs:
            self.fix_arg(arg, parent, ctx)

    def fix_arg(self, node: ast.arg, parent: ast.AST, ctx: ast.AST) -> None:
        if isinstance(ctx, ast.ClassDef) and node.arg in (NAME_CLS, NAME_SELF):
            return

        if not node.annotation:
            node.annotation = ast.Name(id=NAME_ANY)
            self.import_tracker.add_import("Any", "typing")
            self.change_tracker.add_change("missing-arg-type", self.get_range(node), node, parent)

    def get_range(self, node: ast.arg) -> tuple[int, int]:
        lineno = node.lineno
        end_lineno = node.end_lineno or node.lineno
        return (lineno, end_lineno)


class ClassFixer(Fixer):
    METHODS_MAP = [
        ("None", {"__init__"}),
        ("bool", {"__lt__", "__le__", "__ne__", "__eq__", "__gt__", "__ge__", "__bool__"}),
        ("int", {"__hash__"}),
        ("Self", {"__new__"}),
        ("str", {"__str__", "__repr__"}),
    ]
    UNKNOWN_RETURN: list[ast.Return | ast.Yield | ast.YieldFrom] = []

    @classmethod
    def get_fix_types(cls) -> FixTypes:
        return [ast.FunctionDef, ast.AsyncFunctionDef]

    def fix(self, node: ast.AST, _: ast.AST, ctx: ast.AST) -> None:
        if isinstance(node, ast.FunctionDef | ast.AsyncFunctionDef):
            self.fix_func(node, ctx)

    def fix_func(self, node: ast.FunctionDef | ast.AsyncFunctionDef, ctx: ast.AST) -> None:
        if isinstance(ctx, ast.ClassDef):
            for name, methods in self.METHODS_MAP:
                if node.name in methods and not node.returns:
                    if name == "Self":
                        node.returns = ast.Name(id=ast.Constant(ctx.name), ctx=node)
                    else:
                        node.returns = ast.Name(id=name, ctx=node)
                    range = self.get_range(node)
                    self.import_tracker.add_import("Any", "typing")
                    self.change_tracker.add_change(
                        f"missing-return-type-{name.lower()}", range, node, node
                    )

        if not node.returns:
            node.returns = self.guess_return_type(node, ctx)
            range = self.get_range(node)
            self.import_tracker.add_import("Any", "typing")
            self.change_tracker.add_change("missing-return-type-any", range, node, node)

    def get_range(self, node: ast.FunctionDef | ast.AsyncFunctionDef) -> tuple[int, int]:
        lineno = node.lineno
        end_lineno = node.end_lineno or node.lineno
        if node.body:
            end_lineno = node.body[0].lineno
        return (lineno, end_lineno)

    def guess_return_type(
        self,
        node: ast.FunctionDef | ast.AsyncFunctionDef,
        ctx: ast.AST,
    ) -> ast.expr:
        returns = self.get_returns(node)
        if returns is self.UNKNOWN_RETURN:
            return ast.Name(id=NAME_ANY, ctx=node)
        elif len(returns) == 0:
            return ast.Name(id=NAME_NONE, ctx=node)
        elif (  # if only 1 return and it's a call to a class, we can assume it as the type
            len(returns) == 1
            and isinstance(returns[0], ast.Return)
            and isinstance(returns[0].value, ast.Call)
            and isinstance(returns[0].value.func, ast.Name)
            and returns[0].value.func.id[0].isupper()
            and "_" not in returns[0].value.func.id
        ):
            name = returns[0].value.func.id
            if isinstance(ctx, ast.ClassDef) and ctx.name == name:
                return ast.Constant(value=name)
            else:
                return ast.Name(id=name, ctx=node)
        elif len(returns) == 1:
            # print(self.opts.filename)
            # print(returns[0].value.__dict__)
            return ast.Name(id=NAME_ANY, ctx=node)
        else:
            return ast.Name(id=NAME_ANY, ctx=node)

    def get_returns(
        self,
        node: ast.FunctionDef | ast.AsyncFunctionDef,
    ) -> list[ast.Return | ast.Yield | ast.YieldFrom]:
        try:
            visitor = ReturnVisitor()
            visitor.visit(node)
            return visitor.returns
        except RecursionError:
            return self.UNKNOWN_RETURN


class ReturnVisitor(ast.NodeVisitor):
    returns: list[ast.Return | ast.Yield | ast.YieldFrom] = []

    def visit_Return(self, node: ast.Return) -> None:  # noqa: N802
        self.returns.append(node)

    def visit_Yield(self, node: ast.Yield) -> None:  # noqa: N802
        self.returns.append(node)

    def visit_YieldFrom(self, node: ast.YieldFrom) -> None:  # noqa: N802
        self.returns.append(node)
