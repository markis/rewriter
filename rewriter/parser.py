import re
from ast import Module, parse, unparse
from difflib import Differ

from black import format_str
from black.mode import Mode, TargetVersion

from rewriter.fixers import FixerStats
from rewriter.options import Options


def parse_tree(opts: Options) -> Module:
    with open(opts.filename) as f:
        opts.source = f.read()
    return parse(opts.source, opts.filename)


def unparse_tree(opts: Options, tree: Module, stats: FixerStats) -> str:
    if not stats:
        return opts.source

    result = unparse(tree)
    reformatted = reformat(result)
    final = merge_new_code(opts, reformatted, stats)

    if not opts.dry_run:
        with open(opts.filename, "w") as f:
            f.write(final)
    return final


def merge_new_code(opts: Options, result: str, stats: FixerStats) -> str:
    result_lines = result.strip().splitlines()
    original = opts.source.splitlines()

    lineno = 0
    final: list[str] = []
    ranges = get_ranges(stats)
    differ = Differ()
    for diff in differ.compare(original, result_lines):
        diff_type = diff[0]
        line = diff[2:]

        if diff_type in (" ", "-"):
            lineno += 1
        in_range = lineno in ranges

        if diff_type == " ":
            final.append(f"{line}\n")

        elif diff_type == "-" and not in_range:
            final.append(f"{line}\n")

        elif diff_type == "+" and in_range:
            final.append(f"{line}\n")

        if opts.verbose:
            indicator = "*" if in_range else " "
            print(f"{lineno} {indicator}: {diff_type} {line}")

    final = add_any_import(final)

    return "".join(final)


def add_any_import(lines: list[str]) -> list[str]:
    # TODO: This is "good enough" to add the Any import, but this should be revisited

    any_check = re.compile("from typing import .*Any")
    imports_start = 0
    for idx, line in enumerate(lines):
        if line.startswith("from") or line.startswith("import"):
            imports_start = idx
        if any_check.match(line):
            return lines

    lines.insert(imports_start, "from typing import Any\n")
    return lines


def get_ranges(stats: FixerStats) -> set[int]:
    ranges: set[int] = set()
    for stat in stats:
        start, end = stat.range
        for i in range(start, end):
            ranges.add(i)
        else:
            ranges.add(start)
    return ranges


def is_comment(line: str) -> bool:
    return line.lstrip().startswith("#")


def reformat(code: str) -> str:
    black_mode = Mode(target_versions={TargetVersion.PY37}, line_length=100)
    return format_str(code, mode=black_mode)
