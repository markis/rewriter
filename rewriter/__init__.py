import sys
from argparse import ArgumentParser
from pathlib import Path
from typing import Any

from rewriter.options import Options
from rewriter.parser import parse_tree, unparse_tree
from rewriter.trackers.changes import ChangeTracker
from rewriter.trackers.imports import ImportTracker
from rewriter.transformers import Transformer
from rewriter.walker import Walker

__all__ = [
    "Transformer",
    "Options",
    "Walker",
    "ImportTracker",
    "ChangeTracker",
    "parse_tree",
    "unparse_tree",
]

COMPILED = Path(__file__).suffix in (".pyd", ".so")


def parse_options(args: Any = sys.argv[1:]) -> Options:
    is_compiled = "yes" if COMPILED else "no"
    parser = ArgumentParser(
        prog="rewriter", description=f"Rewrite python code (compiled: {is_compiled})"
    )
    parser.add_argument("filename", help="file to rewrite")
    parser.add_argument(
        "-d",
        "--dry-run",
        action="store_true",
        help="execute rewrite but don't update the original file",
    )
    parser.add_argument("-v", "--verbose", action="store_true", help="print diff changes")

    parsed_args: Any = parser.parse_args(args)
    opts: Options = parsed_args
    return opts


def main() -> None:
    opts = parse_options()
    tree = parse_tree(opts)

    changes = ChangeTracker()
    imports = ImportTracker()
    walker = Walker(opts, changes, imports)
    walker.walk(tree)

    unparse_tree(opts, tree, changes, imports)


if __name__ == "__main__":
    main()
