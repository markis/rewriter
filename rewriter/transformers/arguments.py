import ast

from rewriter.trackers.changes import Change
from rewriter.trackers.imports import Import
from rewriter.transformers import TransformerReturnType, transformer

NAME_ANY = "Any"
NAME_CLS = "cls"
NAME_SELF = "self"


@transformer(ast.arguments)
def transform_args(node: ast.AST, parent: ast.AST, ctx: ast.AST) -> TransformerReturnType:
    changes: list[Change] = []
    imports: list[Import] = []

    if not isinstance(node, ast.arguments):
        return changes, imports

    def get_range(node: ast.arg) -> tuple[int, int]:
        lineno = node.lineno
        end_lineno = node.end_lineno or node.lineno
        return (lineno, end_lineno)

    def transform_arg(node: ast.arg, _: ast.AST, ctx: ast.AST) -> None:
        if isinstance(ctx, ast.ClassDef) and node.arg in (NAME_CLS, NAME_SELF):
            return

        if not node.annotation:
            node.annotation = ast.Name(id=NAME_ANY)
            imports.append(Import("typing", "Any"))
            changes.append(Change(get_range(node), "missing-arg-type"))

    if node.vararg:
        transform_arg(node.vararg, parent, ctx)

    if node.kwarg:
        transform_arg(node.kwarg, parent, ctx)

    for arg in node.args:
        transform_arg(arg, parent, ctx)

    for arg in node.posonlyargs:
        transform_arg(arg, parent, ctx)

    for arg in node.kwonlyargs:
        transform_arg(arg, parent, ctx)

    return changes, imports
