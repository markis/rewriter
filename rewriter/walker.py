import ast
from collections.abc import Sequence
from typing import overload

from rewriter.trackers.changes import Change
from rewriter.trackers.imports import Import
from rewriter.transformers import Transform, TransformerReturnType

WalkableNodeType = ast.Module | ast.AST | Sequence[ast.AST]
EMPTY_AST: ast.AST = ast.AST()


def walk(tree: ast.Module, transform: Transform) -> TransformerReturnType:
    changes: list[Change] = []
    imports: list[Import] = []

    @overload
    def walker(node: ast.Module) -> None:
        ...

    @overload
    def walker(node: ast.AST, parent: ast.AST, ctx: ast.AST) -> None:
        ...

    @overload
    def walker(node: Sequence[ast.AST], parent: ast.AST, ctx: ast.AST) -> None:
        ...

    def walker(
        node: WalkableNodeType, parent: ast.AST = EMPTY_AST, ctx: ast.AST = EMPTY_AST
    ) -> None:
        if isinstance(node, Sequence):
            for n in node:
                walker(n, parent, ctx)
        else:
            if isinstance(node, ast.Module | ast.ClassDef):
                walker(node.body, node, node)
            elif isinstance(node, ast.FunctionDef | ast.AsyncFunctionDef):
                walker(node.body, node, ctx)
                walker(node.args, node, ctx)
            elif hasattr(node, "body"):
                walker(getattr(node, "body"), parent, ctx)
            elif hasattr(node, "args"):
                walker(getattr(node, "args"), parent, ctx)

            for new_changes, new_imports in transform(node, parent, ctx):
                changes.extend(new_changes)
                imports.extend(new_imports)

    walker(tree)
    return changes, imports
