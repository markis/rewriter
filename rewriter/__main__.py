from rewriter.options import parse_options
from rewriter.parser import parse_tree, unparse_tree
from rewriter.walker import Walker

opts = parse_options()
tree = parse_tree(opts)

walker = Walker(opts, tree)
walker.walk()

unparse_tree(opts, tree, walker.stats)
