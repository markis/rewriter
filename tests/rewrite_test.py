import ast
from textwrap import dedent
from unittest.mock import Mock

from syrupy.assertion import SnapshotAssertion

from rewriter.options import Options
from rewriter.parser import unparse_tree
from rewriter.transformers import transform
from rewriter.walker import walk


def rewrite_code(source: str) -> str:
    source = dedent(source).lstrip()
    mock_opts = Mock(spec=Options, filename="test.py", source=source, dry_run=True, verbose=True)
    tree = ast.parse(source)
    changes, imports = walk(transform, tree)
    return unparse_tree(mock_opts, tree, changes, imports)


def test_func_definition(snapshot: SnapshotAssertion) -> None:
    source = """
    def test(x, *args, **kwargs):
        pass
    """

    assert rewrite_code(source) == snapshot


def test_class_definition(snapshot: SnapshotAssertion) -> None:
    source = """
    class Test:
        def __init__(self, *, x):
            pass

        def do(self, x):
            return 1 / 2
    """

    assert rewrite_code(source) == snapshot
