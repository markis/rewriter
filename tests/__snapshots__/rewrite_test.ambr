# serializer version: 1
# name: test_class_definition
  '''
  class Test:
      def __init__(self, *, x: Any) -> None:
          pass
  
      def do(self, x: Any) -> Any:
          return 1 / 2
  
  '''
# ---
# name: test_func_definition
  '''
  from typing import Any
  
  
  def test(x: Any, *args: Any, **kwargs: Any) -> None:
      pass
  
  '''
# ---
